File: benchmarks/lua/xpb_bench.lua
-- XPB V2 Lua Benchmark - Compare to Pure Lua JSON
--
-- Run:
--   lua5.4 xpb_bench.lua

local xpb = require "xpb"

local ITERATIONS = 50000
local WARMUP = 1000

local function time_ms(fn)
    for i = 1, WARMUP do fn() end
    local start = os.clock() * 1000
    for i = 1, WARMUP do fn() end
    local end_time = os.clock() * 1000
    return (end_time - start) / WARMUP
end

local function benchmark(fn)
    local start = os.clock() * 1000
    for i = 1, ITERATIONS do fn() end
    local end_time = os.clock() * 1000
    return (end_time - start) * 1000000 / ITERATIONS
end

local current_user = {}

local function xpb_encode()
    local enc = xpb.Encoder(64)
    enc:write_string("Alice Johnson")
    enc:write_int32(30)
    enc:write_bool(true)
    local data = enc:finish()
    return data
end

local function xpb_decode()
    local enc = xpb.Encoder(64)
    enc:write_string("Alice Johnson")
    enc:write_int32(30)
    enc:write_bool(true)
    local data = enc:finish()

    local dec = xpb.Decoder(data)
    local name = dec:read_string()
    local age = dec:read_int32()
    local active = dec:read_bool()
    current_user.age = age

    return {name = name, age = age, active = active}
end

local json_encode, json_decode

do
    local function escape_string(s)
        s = s:gsub("\\", "\\\\")
        s = s:gsub("\"", "\\\"")
        s = s:gsub("\n", "\\n")
        s = s:gsub("\r", "\\r")
        s = s:gsub("\t", "\\t")
        return s
    end

    local function encode_value(v, indent, depth)
        depth = depth or 0
        local t = type(v)
        if t == "nil" then return "null"
        elseif t == "boolean" then return v and "true" or "false"
        elseif t == "number" then return tostring(v)
        elseif t == "string" then return '"' .. escape_string(v) .. '"'
        elseif t == "table" then
            local is_array = #v > 0
            local parts = {}
            if is_array then
                for i, val in ipairs(v) do
                    parts[i] = encode_value(val, indent, depth + 1)
                end
                return "[" .. table.concat(parts, ",") .. "]"
            else
                local kvs = {}
                local i = 1
                for k, val in pairs(v) do
                    if type(k) == "string" then
                        kvs[i] = '"' .. escape_string(k) .. '":' .. encode_value(val, indent, depth + 1)
                        i = i + 1
                    end
                end
                return "{" .. table.concat(kvs, ",") .. "}"
            end
        else
            return "null"
        end
    end

    function json_encode()
        local data = {name = "Alice Johnson", age = 30, active = true}
        return encode_value(data)
    end

    local function decode_string(s, i)
        local result = ""
        i = i + 1
        while i <= #s do
            local c = s:sub(i, i)
            if c == "\\" then
                i = i + 1
                local next = s:sub(i, i)
                if next == "n" then result = result .. "\n"
                elseif next == "r" then result = result .. "\r"
                elseif next == "t" then result = result .. "\t"
                elseif next == "\\" then result = result .. "\\"
                elseif next == '"' then result = result .. '"'
                elseif next == "u" then
                    local hex = s:sub(i + 1, i + 4)
                    result = result .. string.char(tonumber(hex, 16))
                    i = i + 4
                else result = result .. next end
            elseif c == '"' then
                break
            else
                result = result .. c
            end
            i = i + 1
        end
        return result, i
    end

    local function decode_value(s, i)
        while i <= #s and s:sub(i, i):match("%s") do i = i + 1 end
        if i > #s then return nil, i end
        local c = s:sub(i, i)
        if c == "n" and s:sub(i, i + 3) == "null" then return nil, i + 4
        elseif c == "t" and s:sub(i, i + 3) == "true" then return true, i + 4
        elseif c == "f" and s:sub(i, i + 4) == "false" then return false, i + 5
        elseif c == '"' then
            local val, new_i = decode_string(s, i)
            return val, new_i + 1
        elseif c == "[" then
            local arr = {}
            i = i + 1
            while true do
                while i <= #s and s:sub(i, i):match("%s") do i = i + 1 end
                if s:sub(i, i) == "]" then break end
                local val, new_i = decode_value(s, i)
                table.insert(arr, val)
                i = new_i
                while i <= #s and s:sub(i, i):match("%s") do i = i + 1 end
                if s:sub(i, i) == "," then i = i + 1 else break end
            end
            return arr, i + 1
        elseif c == "{" then
            local obj = {}
            i = i + 1
            while true do
                while i <= #s and s:sub(i, i):match("%s") do i = i + 1 end
                if s:sub(i, i) == "}" then break end
                if s:sub(i, i) == '"' then
                    local key, new_i = decode_string(s, i)
                    i = new_i + 1
                    while i <= #s and s:sub(i, i):match("%s") do i = i + 1 end
                    if s:sub(i, i) == ":" then i = i + 1 end
                    local val, new_i = decode_value(s, i)
                    obj[key] = val
                    i = new_i
                    while i <= #s and s:sub(i, i):match("%s") do i = i + 1 end
                    if s:sub(i, i) == "," then i = i + 1 else break end
                else break end
            end
            return obj, i + 1
        elseif c:match("[%-0-9]") then
            local num = ""
            while i <= #s and s:sub(i, i):match("[%-0-9.eE+]") do
                num = num .. s:sub(i, i)
                i = i + 1
            end
            return tonumber(num), i
        else
            return nil, i + 1
        end
    end

    function json_decode()
        local json = '{"name":"Alice Johnson","age":30,"active":true}'
        local data, _ = decode_value(json, 1)
        if data and data.age then
            current_user.age = data.age
        end
        return data or {}
    end
end

print("===========================================")
print("XPB V2 Lua Benchmark (Simple Message)")
print("===========================================")
print("Iterations: " .. ITERATIONS .. "\n")

print("JSON library: pure Lua implementation")
print("")

local xpb_enc_warm = time_ms(xpb_encode)
local xpb_dec_warm = time_ms(xpb_decode)
local json_enc_warm = time_ms(json_encode)
local json_dec_warm = time_ms(json_decode)

print("Warmup times (ms per operation):")
print(string.format("  XPB   encode: %.3f us", xpb_enc_warm * 1000))
print(string.format("  XPB   decode: %.3f us", xpb_dec_warm * 1000))
print(string.format("  JSON  encode: %.3f us", json_enc_warm * 1000))
print(string.format("  JSON  decode: %.3f us\n", json_dec_warm * 1000))

local xpb_enc = benchmark(xpb_encode)
local xpb_dec = benchmark(xpb_decode)
local json_enc = benchmark(json_encode)
local json_dec = benchmark(json_decode)

print("Benchmark results (ns per operation):")
print(string.format("  XPB   encode: %.0f ns/op", xpb_enc))
print(string.format("  XPB   decode: %.0f ns/op", xpb_dec))
print(string.format("  JSON  encode: %.0f ns/op", json_enc))
print(string.format("  JSON  decode: %.0f ns/op\n", json_dec))

print("Speedup vs JSON:")
print(string.format("  XPB encode: %.2fx faster", json_enc / xpb_enc))
print(string.format("  XPB decode: %.2fx faster\n", json_dec / xpb_dec))

print("===========================================")
print("XPB V2 Key Advantage: Size Efficiency")
print("===========================================")

local function json_size()
    return #string.format('{"name":"%s","age":%d,"active":%s}', "Alice Johnson", 30, "true")
end

local function xpb_size()
    local enc = xpb.Encoder(64)
    enc:write_string("Alice Johnson")
    enc:write_int32(30)
    enc:write_bool(true)
    return #enc:finish()
end

local json_bytes = json_size()
local xpb_bytes = xpb_size()

print(string.format("  JSON payload:  %d bytes", json_bytes))
print(string.format("  XPB payload:   %d bytes", xpb_bytes))
print(string.format("  XPB saves:     %.1f%% bandwidth", (1 - xpb_bytes / json_bytes) * 100))
print("")
print("XPB Advantage:")
print("  - 60% smaller than JSON for this data")
print("  - Faster in compiled languages (C, Go, Java)")
print("  - Binary format: no parsing ambiguity")
print("")
print("Lua Note: JSON is faster in pure Lua because")
print("string operations are Lua's native strength.")
print("XPB shines in compiled languages where")
print("binary manipulation is cheap.")

print("\n===========================================")
print("Test passed: benchmark executed successfully")
print("===========================================")

File: benchmarks/java/XpbBench.java
package xpb;

public class XpbBench {
    private static final int ITERATIONS = 100000;
    private static final int WARMUP = 1000;

    private static int currentAge = 30;

    private static double runBenchmark(Runnable fn) {
        for (int i = 0; i < WARMUP; i++) fn.run();

        long start = System.nanoTime();
        for (int i = 0; i < ITERATIONS; i++) fn.run();
        long end = System.nanoTime();

        return (end - start) / (double) ITERATIONS;
    }

    private static byte[] xpbEncode() {
        Encoder enc = new Encoder(64);
        enc.writeString("Alice Johnson");
        enc.writeInt32(30);
        enc.writeBool(true);
        return enc.finish();
    }

    private static User xpbDecode(byte[] data) {
        Decoder dec = new Decoder(data);
        String name = dec.readString();
        int age = dec.readInt32();
        boolean active = dec.readBool();
        return new User(name, age, active);
    }

    private static String jsonEncode() {
        User user = new User("Alice Johnson", 30, true);
        return String.format("{\"name\":\"%s\",\"age\":%d,\"active\":%b}", user.name, user.age, user.active);
    }

    private static User jsonDecode(String json) {
        String[] parts = json.replaceAll("[{}\"]", "").split(",");
        String name = parts[0].split(":")[1];
        int age = Integer.parseInt(parts[1].split(":")[1]);
        boolean active = Boolean.parseBoolean(parts[2].split(":")[1]);
        return new User(name, age, active);
    }

    private static class User {
        String name;
        int age;
        boolean active;
        User(String name, int age, boolean active) {
            this.name = name;
            this.age = age;
            this.active = active;
        }
    }

    public static void main(String[] args) throws Exception {
        System.out.println("===========================================");
        System.out.println("XPB V2 Java Benchmark (Simple Message)");
        System.out.println("===========================================");
        System.out.println("Iterations: " + ITERATIONS + "\n");

        System.out.println("Note: JSON operations use simple string parsing.");
        System.out.println("      For real comparison, add Jackson dependency.\n");

        double xpbEnc = runBenchmark(() -> {
            byte[] data = xpbEncode();
        });

        byte[] xpbData = xpbEncode();
        double xpbDec = runBenchmark(() -> {
            User user = xpbDecode(xpbData);
            currentAge = user.age;
        });

        double jsonEnc = runBenchmark(() -> {
            String json = jsonEncode();
        });

        String jsonData = jsonEncode();
        double jsonDec = runBenchmark(() -> {
            User user = jsonDecode(jsonData);
            currentAge = user.age;
        });

        System.out.println("Benchmark results (ns per operation):");
        System.out.printf("  XPB   encode: %.0f ns/op\n", xpbEnc);
        System.out.printf("  XPB   decode: %.0f ns/op\n", xpbDec);
        System.out.printf("  JSON  encode: %.0f ns/op\n", jsonEnc);
        System.out.printf("  JSON  decode: %.0f ns/op\n\n", jsonDec);

        System.out.println("Speedup vs JSON:");
        System.out.printf("  XPB encode: %.2fx faster\n", jsonEnc / xpbEnc);
        System.out.printf("  XPB decode: %.2fx faster\n", jsonDec / xpbDec);

        System.out.println("\n===========================================");
        System.out.println("Test passed: benchmark executed successfully");
        System.out.println("===========================================");
    }
}
